### PR TITLE
Map the stack guard page with max protection on NetBSD

### LINKED ISSUE #50313: Boostraped rustc fails in thread initialization on NetBSD
Trying to build rustc on NetBSD fails, after creating a local rustc, it fails when invoked for the first time:

thread 'main' panicked at 'unable to reset the guard page',libstd/sys/unix/thread.rs:361:21

I have no idea why that mprotect() call fails, it would be helpfull if the panic would mention the address and the return value.

Unfortunately I can't just invoke the same binary manually as it needs some secret/hidden environment setup to work.

### PR DESCRIPTION
**Pull request by MartinHusemann**
On NetBSD the initial mmap() protection of a mapping can not be made
less restrictive with mprotect().

So when mapping a stack guard page, use the maximum protection
we ever want to use, then mprotect() it to the permission we
want it to have initially.

Fixes #50313

### CODE REVIEW

File: src/libstd/sys/unix/thread.rs [hunk 1]
```diff
@@ -326,11 +326,23 @@ pub mod guard {
             // Reallocate the last page of the stack.
             // This ensures SIGBUS will be raised on
             // stack overflow.
-            let result = mmap(stackaddr, PAGE_SIZE, PROT_NONE,
-                              MAP_PRIVATE | MAP_ANON | MAP_FIXED, -1, 0);
+            if cfg!(target_os = "netbsd") {
```
Comment (estebank) [inline src/libstd/sys/unix/thread.rs]: Can you add a comment with an abridged version of the PR description, mentioning the Issue #? That way people can track the reason for this without having to use `git blame` to fish out why netbsd is being singled out here.

**Review by estebank — COMMENTED** [review]
Does this affect _only_ NetBSD?

File: src/libstd/sys/unix/thread.rs (hunk 1 above)
Comment (zoulasc) [reply src/libstd/sys/unix/thread.rs]: It affects anyone who is enforcing PaX MPROTECT in strict mode. Strict mode means not only you are enforcing W^X (i.e. you can't map a segment both writable and executable at the same time), but also that you can't use mprotect(2) to augment (add) permissions on an already mapped page (that did not have those permissions during mapping time). I.e. can't toggle between R-X<->RW- using mprotect(2). I am not aware of other OS's doing this at this time, but the change to do the mmap(PROT_READ|PROT_WRITE)+mprotect(PROT_NONE) for the guard page should work on every OS, so it doesn't really need to be conditionalized. I would put a comment why this is the case though (why we are going doing this).

**Comment by shepmaster** [comment]
Ping from triage, @Kimundi  / @estebank — this is ready for re-review.

**Comment by pietroalbini** [comment]
Ping from triage! This PR needs a review, can @Kimundi or someone else from @rust-lang/libs review this?

**Pull request merged**